File: front/src/services/notificationService.ts
import { apiService } from './api';

export interface NotificationDTO {
  id: number;
  userId: number;
  title: string;
  message: string;
  type: 'INFO' | 'SUCCESS' | 'WARNING' | 'ERROR';
  isRead: boolean;
  link?: string;
  createdAt: string;
}

export const notificationService = {
  async getUserNotifications(userId: number): Promise<NotificationDTO[]> {
    return apiService.get<NotificationDTO[]>(`/notifications/user/${userId}`);
  },

  async getUnreadNotifications(userId: number): Promise<NotificationDTO[]> {
    return apiService.get<NotificationDTO[]>(`/notifications/user/${userId}/unread`);
  },

  async getUnreadCount(userId: number): Promise<number> {
    return apiService.get<number>(`/notifications/user/${userId}/unread/count`);
  },

  async markAsRead(notificationId: number): Promise<void> {
    return apiService.patch<void>(`/notifications/${notificationId}/read`, {});
  },

  async markAllAsRead(userId: number): Promise<void> {
    return apiService.patch<void>(`/notifications/user/${userId}/read-all`, {});
  },

  async deleteNotification(notificationId: number): Promise<void> {
    return apiService.delete<void>(`/notifications/${notificationId}`);
  }
};

File: src/main/java/com/internship/management/service/NotificationService.java
package com.internship.management.service;

import com.internship.management.dto.NotificationDTO;
import com.internship.management.entity.Notification;
import com.internship.management.entity.User;
import com.internship.management.repository.NotificationRepository;
import com.internship.management.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class NotificationService {

    private final NotificationRepository notificationRepository;
    private final UserRepository userRepository;

    @Transactional(readOnly = true)
    public List<NotificationDTO> getUserNotifications(Long userId) {
        return notificationRepository.findByUserIdOrderByCreatedAtDesc(userId).stream()
                .map(NotificationDTO::fromEntity)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<NotificationDTO> getUnreadNotifications(Long userId) {
        return notificationRepository.findByUserIdAndIsReadOrderByCreatedAtDesc(userId, false).stream()
                .map(NotificationDTO::fromEntity)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public Long getUnreadCount(Long userId) {
        return notificationRepository.countByUserIdAndIsRead(userId, false);
    }

    @Transactional
    public NotificationDTO createNotification(NotificationDTO notificationDTO) {
        User user = userRepository.findById(notificationDTO.getUserId())
                .orElseThrow(() -> new RuntimeException("Utilisateur introuvable"));

        Notification notification = Notification.builder()
                .user(user)
                .title(notificationDTO.getTitle())
                .message(notificationDTO.getMessage())
                .type(Notification.NotificationType.valueOf(notificationDTO.getType()))
                .isRead(false)
                .link(notificationDTO.getLink())
                .build();

        Notification saved = notificationRepository.save(notification);
        return NotificationDTO.fromEntity(saved);
    }

    @Transactional
    public void markAsRead(Long notificationId) {
        Notification notification = notificationRepository.findById(notificationId)
                .orElseThrow(() -> new RuntimeException("Notification introuvable"));
        notification.setIsRead(true);
        notificationRepository.save(notification);
    }

    @Transactional
    public void markAllAsRead(Long userId) {
        List<Notification> notifications = notificationRepository.findByUserIdAndIsReadOrderByCreatedAtDesc(userId, false);
        notifications.forEach(n -> n.setIsRead(true));
        notificationRepository.saveAll(notifications);
    }

    @Transactional
    public void deleteNotification(Long notificationId) {
        notificationRepository.deleteById(notificationId);
    }
}
